libazureinit/src/status.rs

```diff
@@ -0,0 +1,342 @@
+                let swapped_uuid = Uuid::from_bytes(swapped_bytes);
+                let final_id = swapped_uuid.to_string();
+                tracing::info!(target: "libazureinit::status::retrieved_vm_id", "VM ID (Gen1, swapped): {}", final_id);
+                Some(final_id)
+            }
+            Err(e) => {
+                tracing::error!(
+                    "Failed to parse system UUID '{}': {}",
+                    system_uuid,
+                    e
+                );
+                // fallback to the raw string
+                Some(system_uuid)
+            }
+        }
+    } else {
+        tracing::info!(target: "libazureinit::status::retrieved_vm_id", "VM ID (Gen2, no swap): {}", system_uuid);
+        Some(system_uuid)
+    }
+}
+
+/// This function checks whether a provisioning status file exists for the current VM ID.
+/// If the file exists, provisioning has already been completed and should be skipped.
+/// If the file does not exist or the VM ID has changed, provisioning should proceed.
+///
+/// - Returns `true` if provisioning is complete (i.e., provisioning file exists).
+/// - Returns `false` if provisioning has not been completed (i.e., no provisioning file exists).
+pub fn is_provisioning_complete(
+    config: Option<&Config>,
+    vm_id: Option<String>,
+) -> bool {
+    let vm_id = vm_id.or_else(|| get_vm_id(None, None));
+
+    if let Some(vm_id) = vm_id {
+        let file_path =
+            get_provisioning_dir(config).join(format!("{}.provisioned", vm_id));
+        if std::path::Path::new(&file_path).exists() {
+            tracing::info!("Provisioning already complete. Skipping...");
+            return true;
+        }
+    }
+    tracing::info!("Provisioning required.");
+    false
+}
+
+/// This function creates an empty file named after the current VM ID in the
+/// provisioning directory. The presence of this file signals that provisioning
+/// has been successfully completed.
+pub fn mark_provisioning_complete(
+    config: Option<&Config>,
+    vm_id: Option<String>,
+) -> Result<(), Error> {
+    check_provision_dir(config)?;
+
+    let vm_id = vm_id.or_else(|| get_vm_id(None, None));
+
+    if let Some(vm_id) = vm_id {
+        let file_path =
+            get_provisioning_dir(config).join(format!("{}.provisioned", vm_id));
+
+        if let Err(error) = File::create(&file_path) {
+            tracing::error!(
+                ?error,
+                file_path=?file_path,
+                "Failed to create provisioning status file"
+            );
+            return Err(error.into());
+        }
+
+        tracing::info!(
+            target: "libazureinit::status::success",
+            "Provisioning complete. File created: {}",
+            file_path.display()
+        );
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+    use tempfile::TempDir;
+
+    #[test]
+    fn test_mark_provisioning_complete() {
```

Comment:
I'm seeing this test fail occasionally locally. I believe this is because the cargo test runner is multi-threaded, so all your tests are stepping on top of each other. It's also worth noting that, unfortunately, setting environment variables in *nix is not [thread safe](https://doc.rust-lang.org/std/env/fn.set_var.html#safety). 

It's best to avoid setting an environment variable from your program if you can possibly help it. Even if you're absolutely sure there's no other threads, someone might come along in the future and add multi-threading without noticing the environment is getting modified.

Reply:
I believe this was addressed, resolving.